contracts/CompoundFCM.sol

```diff
@@ -0,0 +1,337 @@
+    // uint256 currentExchangeRate = _ctoken.exchangeRateStored();
+    // marginInYieldBearingTokens = traderMarginInScaledYieldBearingTokens.rayMul(currentExchangeRate);
+    return traderMarginInScaledYieldBearingTokens; // NO scaling to do here. Delete this function?
+  }
+
+  function getTraderMarginInCTokens(address traderAddress)
+        external
+        view
+        returns (uint256 marginInYieldBearingTokens)
+    {
+        TraderWithYieldBearingAssets.Info storage trader = traders[
+            traderAddress
+        ];
+        marginInYieldBearingTokens = trader.marginInScaledYieldBearingTokens;
+    }
+
+
+  /// @notice Unwind Fully Collateralised Fixed Taker Swap
+  /// @param notionalToUnwind The amount of notional to unwind (stop securing with a fixed rate)
+  /// @param sqrtPriceLimitX96 The sqrt price limit (binary fixed point notation) beyond which the unwind cannot progress
+  function unwindFullyCollateralisedFixedTakerSwap(uint256 notionalToUnwind, uint160 sqrtPriceLimitX96) external override {
+
+    // add require statement and isApproval
+
+    TraderWithYieldBearingAssets.Info storage trader = traders[msg.sender];
+
+    require(trader.variableTokenBalance <= 0, "Trader VT balance positive");
+
+    /// @dev it is impossible to unwind more variable token exposure than the user already has
+    /// @dev hencel, the notionalToUnwind needs to be <= absolute value of the variable token balance of the trader
+    /// @audit-casting variableTokenDelta is expected to be negative here, but what if goes above 0 due to rounding imprecision?
+    require(uint256(-trader.variableTokenBalance) >= notionalToUnwind, "notional to unwind > notional");
+
+    // initiate a swap
+    /// @dev as convention, specify the tickLower to be equal to -tickSpacing and tickUpper to be equal to tickSpacing
+    // since the unwind is in the Variable Taker direction, the amountSpecified needs to be exact output => needs to be negative = -int256(notionalToUnwind),
+    IVAMM.SwapParams memory params = IVAMM.SwapParams({
+        recipient: address(this),
+        amountSpecified: -notionalToUnwind.toInt256(),
+        sqrtPriceLimitX96: sqrtPriceLimitX96,
+        tickLower: -tickSpacing,
+        tickUpper: tickSpacing
+    });
+
+    (int256 fixedTokenDelta, int256 variableTokenDelta, uint256 cumulativeFeeIncurred, ,) = _vamm.swap(params);
+
+    require(variableTokenDelta >= 0, "VT delta negative");
+
+    // update trader fixed and variable token balances
+    (int256 _fixedTokenBalance, int256 _variableTokenBalance) = trader.updateBalancesViaDeltas(fixedTokenDelta, variableTokenDelta);
+
+    uint256 currentExchangeRate = _ctoken.exchangeRateStored();
+
+    // TODO: does this work with arbitrary numbers of decimals in the exchange rate (exchange rate decimals depends on underlying decimals). Probably not?
```

Comment:
need to discuss this

Reply:
It does work. (Though it took me some time to convince myself of that - see https://docs.google.com/spreadsheets/d/1tX8U4yX0ib5aO9WJFgzLipN_U_WG2HwpN77Ndgx8arI/edit?usp=sharing)